shell/platform/darwin/macos/framework/Source/FlutterBackingStoreDescriptor.h

```diff
@@ -0,0 +1,50 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <Cocoa/Cocoa.h>
+#import <Metal/Metal.h>
+
+/**
+ * The type of the backing store, either Metal or OpenGL.
+ */
+typedef NS_ENUM(NSInteger, FlutterMacOSBackingStoreType) {
+  FlutterMacOSBackingStoreTypeMetal,
+  FlutterMacOSBackingStoreTypeOpenGL
+};
+
+/**
+ * A wrapper around backing store handles. On OpenGL, refers to the frame buffer id, and on Metal,
+ * holds the reference to the MTLTexture. Will only contain one of Metal or OpenGL backing store
+ * handles.
+ */
+@interface FlutterBackingStoreDescriptor : NSObject
+
+/**
+ * The type of the backing store, either Metal or OpenGL. See: FlutterMacOSBackingStoreType.
+ */
+@property(readonly, nonatomic) FlutterMacOSBackingStoreType backingStoreType;
+
+/**
+ * Initializes a backing store descriptor with the specified frame buffer id.
+ */
+- (nullable instancetype)initOpenGLDescriptorWithFBOId:(uint32_t)fboId;
+
+/**
+ * Frame buffer ID referenced by this backing store instance. Only valid when `backingStoreType` is
+ * OpenGL.
+ */
+- (uint32_t)frameBufferId;
+
+/**
+ * Initializes a backing store descriptor with the specified MTLTexture.
+ */
+- (nullable instancetype)initMetalDescriptorWithTexture:(nonnull id<MTLTexture>)texture;
+
+/**
+ * MTLTexture referenced by this backing store instance. Only valid when `backingStoreType` is
```

Comment:
Or, actually: given that half the API surface is only valid for one type anyway, which isn't this being done via subclasses? Then you don't need an enum-based pseudo-type system, and comments saying which methods you can and can't call.

Reply:
I wasn't able to come up with one that fitted well. These backing are specific to the rendering type used. Incase of OpenGL, we wish to access the FBO ID and in case of metal we will need to access the metal texture. If we had a type say `FlutterBackingStore`, this would somehow allow for embedder to access `frameBufferId` in-case of OpenGL or texture for metal. See: `fboForFrameInfo` in `FlutterOpenGLRenderer` for example. We could have a subclass and cast it to that, but I wasn't sure if that was better.

Reply:
> I wasn't able to come up with one that fitted well. These backing are specific to the rendering type used.

To clarify, I'm not suggesting you have a fully unified interface, just that you have separate classes that you can put the type-specific methods on.

> We could have a subclass and cast it to that, but I wasn't sure if that was better.

I don't see any reason why checking subclass type and casting would be worse than having an enum and checking the enum value; they are doing very similar things, but the former is already built into the language, while the latter is a custom, secondary type-like system. Using types seems both more idiomatic, and safer (since the compiler will help avoid mistakes as long as you don't do incorrect casting).